<a href="https://colab.research.google.com/github/swamhtetg90/DAVI-CA2/blob/Ben/data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DAVI Data Cleaning
## Group-10
## Group Member: Devendran Yoheswaran, Kaung Myat San, Swam Htet Aung
---

## Meta Data
---
### Student Profiles

> **Note:** Data is manually entered, so the values are not standardized.

| **Field Name**                            | **Description**                                                                                                                          | **Example**                |
| ----------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- | -------------------------- |
| **STUDENT ID**                            | Student ID is made up of three attributes: `<Course code>-<Intake No>/<Index Number of student in the intake>`                           | `1101-013/001`             |
| **GENDER**                                | Gender                                                                                                                                   | `M`, `F`                   |
| **SG CITIZEN**                            | Singapore Citizen                                                                                                                        | `Y` or blank               |
| **SG PR**                                 | Singapore Permanent Resident                                                                                                             | `Y` or blank               |
| **FOREIGNER**                             | Neither SG Citizen nor SG PR (mutually exclusive with SG CITIZEN and SG PR)                                                              | `Y` or blank               |
| **COUNTRY OF OTHER NATIONALITY**          | Country of nationality (only for SG PR or foreigner)                                                                                     | `Malaysia`, `India`, etc.  |
| **DOB**                                   | Date of Birth. Format: `DD/MM/YYYY`                                                                                                      | `04/03/1978`               |
| **HIGHEST QUALIFICATION**                 | Highest qualification attained prior to this course                                                                                      | `Certificate`, `Diploma`   |
| **NAME OF QUALIFICATION AND INSTITUTION** | Institute where the highest qualification was attained                                                                                   | As provided by participant |
| **DATE ATTAINED HIGHEST QUALIFICATION**   | Date when the qualification was awarded. Format: `DD/MM/YYYY`                                                                            | `06/11/2016`               |
| **DESIGNATION**                           | Job designation                                                                                                                          | As provided by participant |
| **COMMENCEMENT DATE**                     | Course start date. Format: `DD/MM/YYYY`                                                                                                  | `06/01/2023`               |
| **COMPLETION DATE**                       | Course end date. Blank if course is ongoing. Format: `DD/MM/YYYY`                                                                        | `06/04/2024`               |
| **FULL-TIME OR PART-TIME**                | Whether the course is Full-time or Part-time                                                                                             | `Full-Time`, `Part-Time`   |
| **COURSE FUNDING**                        | Course funding type:<br>- `Individual`<br>- `Individual - SFC` (SkillsFuture Credit)<br>- `Sponsored`<br>- `Individual - waived App Fee` | `Individual - SFC`         |
| **REGISTRATION FEE**                      | Registration fee in SGD                                                                                                                  | As entered                 |
| **PAYMENT MODE**                          | Mode of payment                                                                                                                          | `NETS`, `Giro`, `PayNow`   |
| **COURSE FEE**                            | Course fee in SGD                                                                                                                        | As entered                 |

---

### Course Codes

| **S/N**         | **Description**        | **Example**                          |
| --------------- | ---------------------- | ------------------------------------ |
| **CODE**        | Course code (4 digits) | `1101`                               |
| **COURSE NAME** | Course name            | `Diploma in Business Administration` |

---

### Semester Results

| **S/N**        | **Description**                                                                                                           | **Example**    |
| -------------- | ------------------------------------------------------------------------------------------------------------------------- | -------------- |
| **STUDENT ID** | Must match the ID in the Student Profile dataset                                                                          | `1101-013/001` |
| **PERIOD**     | Semester number<br>• Certificate: 1 semester<br>• Diploma: 3 semesters<br>• Master’s: 2 semesters (some exceptions apply) | `1`, `2`, `3`  |
| **GPA**        | GPA for the semester.<br>• Max GPA: 4<br>• Pass GPA: Certificate/Diploma = 2, Master's = 2.3                              | `3.2`          |

---


## Importing Modules
---

In [1]:
import pandas as pd
import numpy as np

## Loading Data
---

#### Course_Code Data

In [2]:
course_code = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Course%20Codes.xlsx?raw=true")
course_code.head()

CODE                         COURSE NAME
0  1101  Diploma in Business Administration
1  1102       Diploma in Business Analytics
2  2101    Certificate in Digital Marketing
3  2102        Certificate in HR Management
4  2013   Certificate in Tourism Management

#### Semester Results Data

In [3]:
semester_results = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Semester%20Results.xlsx?raw=true")
semester_results.head()

STUDENT ID PERIOD  GPA
0  1101-009/001  Sem 1  3.5
1  1101-009/001  Sem 2  3.6
2  1101-009/001  Sem 3  3.7
3  1101-009/002  Sem 1  3.4
4  1101-009/002  Sem 2  3.5

#### Student Profiles Data

In [4]:
student_profiles = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Student%20Profiles.xlsx?raw=true")
student_profiles.head()

STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
0  1101-009/001      F              NaN         Y   
1  1101-009/002      F          Y   NaN       NaN   
2  1101-009/003      F              NaN         Y   
3  1101-009/004      F              NaN         Y   
4  1101-009/005      F          Y   NaN       NaN   

  COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
0                     Malaysia  13/09/1981           Certificate   
1                          NaN  26/07/1979           Certificate   
2                        India  01/02/1990                Degree   
3                  Netherlands  20/04/1976               Diploma   
4                          NaN  25/11/1983               Diploma   

               NAME OF QUALIFICATION AND INSTITUTION  \
0                                                SPM   
1                  Certificate in Office Skills, ITE   
2  Bachelor of Business Administration, Universit...   
3  Office Management Diploma, NCOI Rotterdam, The...   
4  Diploma in Business Admininstration, LCCI Leve...   

  DATE ATTAINED HIGHEST QUALIFICATION                  DESIGNATION  \
0                          2018-01-08         Admin & HR Assistant   
1                          2016-06-08              Admin Assistant   
2                          2015-08-08                            -   
3                          2018-02-08  HR Support / Office Manager   
4                          2015-06-08    Executive, Administration   

     COMMENCEMENT DATE      COMPLETION DATE FULL-TIME OR PART-TIME  \
0  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
1  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
2  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
3  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
4  2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   

   COURSE FUNDING REGISTRATION FEE PAYMENT MODE  COURSE FEE  
0      Individual              107         GIRO        5136  
1  Individual-SFC              107         NETS        5136  
2      Individual              107         NETS        5136  
3      Individual              107         NETS        5136  
4       Sponsored              107         GIRO        4812

## Data Analysis
---

### Course Code

In [5]:
course_code.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CODE         7 non-null      int64 
 1   COURSE NAME  7 non-null      object
dtypes: int64(1), object(1)
memory usage: 244.0+ bytes


In [6]:
course_code.head(10)

CODE                                        COURSE NAME
0  1101                 Diploma in Business Administration
1  1102                      Diploma in Business Analytics
2  2101                   Certificate in Digital Marketing
3  2102                       Certificate in HR Management
4  2013                  Certificate in Tourism Management
5  5112  Specialist Diploma in Business Innovation and ...
6  5113          Specialist Diploma in Intelligent Systems

#### Course Code DataFrame Summary

- The **`Course Code` DataFrame** contains **no null values**.
- It has a total of **7 rows**.
- The data appears to be **cleaned and ready for use**.


### Student Profile
---

#### Student_ID

In [7]:
import pandas as pd

# Step 1: Get unique student IDs from each DataFrame
ids_profiles = set(student_profiles['STUDENT ID'].dropna().unique())
ids_results = set(semester_results['STUDENT ID'].dropna().unique())

# Step 2: Compare sets
if ids_profiles == ids_results:
    print("STUDENT ID columns match exactly in both datasets.")
else:
    print("Mismatch found between STUDENT ID columns.")

    # Extra: Show differences
    only_in_profiles = ids_profiles - ids_results
    only_in_results = ids_results - ids_profiles

    if only_in_profiles:
        print("Student IDs only in student_profiles:")
        print(only_in_profiles)

    if only_in_results:
        print("Student IDs only in semester_results:")
        print(only_in_results)


Mismatch found between STUDENT ID columns.
Student IDs only in student_profiles:
{'5113-009/002', '2101-111/001', '2101-111/003', '2101-111/005', '5113-009/001', '5113-009/003', '5113-009/005', '5113-009/006', '2101-111/004', '2101-111/002', '2101-111/008', '5113-009/007', '5113-009/004', '2101-111/006', '2101-111/007'}
Student IDs only in semester_results:
{'2101-106/002', '5112-007/002', '5112-007/003', '2101-106/003', '5112-007/006', '5112-007/005', '2101-106/004', '5112-007/001', '2101-106/001', '2101-106/005', '5112-007/004'}


In [8]:
# Get sets of Student IDs
ids_profiles = set(student_profiles['STUDENT ID'].dropna().unique())
ids_results = set(semester_results['STUDENT ID'].dropna().unique())

# Find mismatched IDs (present in semester_results but not in student_profiles)
only_in_results = ids_results - ids_profiles

# Filter and print those rows from semester_results
mismatched_rows = semester_results[semester_results['STUDENT ID'].isin(only_in_results)]

print("❌ Mismatched rows from semester_results:")
print(mismatched_rows)


❌ Mismatched rows from semester_results:
       STUDENT ID PERIOD  GPA
116  2101-106/001  Sem 1  2.4
117  2101-106/002  Sem 1  3.1
118  2101-106/003  Sem 1  3.4
119  2101-106/004  Sem 1  2.8
120  2101-106/005  Sem 1  2.3
130  5112-007/001  Sem 1  2.9
131  5112-007/001  Sem 2  3.8
132  5112-007/002  Sem 1  3.2
133  5112-007/002  Sem 2  3.3
134  5112-007/003  Sem 1  3.6
135  5112-007/003  Sem 2  3.7
136  5112-007/004  Sem 1  3.5
137  5112-007/004  Sem 2  2.6
138  5112-007/005  Sem 1  2.1
139  5112-007/005  Sem 2  2.3
140  5112-007/006  Sem 1  3.2
141  5112-007/006  Sem 2  3.1


In [9]:
# From Ben, i dont think we should remove students which are not found in semester_results dataset as when i checked the data,
# i found out it is because they just started the course so it could be useful for finding out no of intake over time
# Look Below

import pandas as pd

# Extract all unique prefixes from semester_results
prefixes = set(semester_results['STUDENT ID'].str.extract(r'^(\d{4}-\d{3})')[0])

# Filter student_profiles to include only rows whose STUDENT ID starts with one of the prefixes
matched_profiles = student_profiles[
    student_profiles['STUDENT ID'].str.extract(r'^(\d{4}-\d{3})')[0].isin(prefixes)
].copy()

# Display result
print("✅ Matched student profiles (by prefix only):")
matched_profiles


✅ Matched student profiles (by prefix only):


STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
0    1101-009/001      F              NaN         Y   
1    1101-009/002      F          Y   NaN       NaN   
2    1101-009/003      F              NaN         Y   
3    1101-009/004      F              NaN         Y   
4    1101-009/005      F          Y   NaN       NaN   
..            ...    ...        ...   ...       ...   
295  5113-008/003      F          Y   NaN       NaN   
296  5113-008/004      F          Y   NaN       NaN   
297  5113-008/005      F          Y   NaN       NaN   
298  5113-008/006      F          Y   NaN       NaN   
299  5113-008/007      F        NaN   NaN         Y   

    COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
0                       Malaysia  13/09/1981           Certificate   
1                            NaN  26/07/1979           Certificate   
2                          India  01/02/1990                Degree   
3                    Netherlands  20/04/1976               Diploma   
4                            NaN  25/11/1983               Diploma   
..                           ...         ...                   ...   
295                          NaN  18/11/1991                Degree   
296                          NaN  29/04/1974                Degree   
297                          NaN  19/10/1981                Degree   
298                          NaN  19/03/1971                Degree   
299                        China  28/11/1998                Degree   

                 NAME OF QUALIFICATION AND INSTITUTION  \
0                                                  SPM   
1                    Certificate in Office Skills, ITE   
2    Bachelor of Business Administration, Universit...   
3    Office Management Diploma, NCOI Rotterdam, The...   
4    Diploma in Business Admininstration, LCCI Leve...   
..                                                 ...   
295  Bachelor of Business (Marketing)/\nRMIT Univer...   
296  Bachelor of Commerce in Management and Marketi...   
297  Bachelor of Arts in Human Resource Management,...   
298  Bachelor of Arts (Sociology)/\nState Universit...   
299  Bachelor of Arts with Honours/\nNational Unive...   

    DATE ATTAINED HIGHEST QUALIFICATION                   DESIGNATION  \
0                            2018-01-08          Admin & HR Assistant   
1                            2016-06-08               Admin Assistant   
2                            2015-08-08                             -   
3                            2018-02-08   HR Support / Office Manager   
4                            2015-06-08     Executive, Administration   
..                                  ...                           ...   
295                          2017-03-21            Regional Recruiter   
296                          2017-02-28        Confidential Assistant   
297                          2018-01-30  Journey Management Team Lead   
298                          1995-05-30   Academy Program Coordinator   
299                          2019-11-30          Management Associate   

       COMMENCEMENT DATE      COMPLETION DATE FULL-TIME OR PART-TIME  \
0    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
1    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
2    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
3    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
4    2022-04-18 00:00:00  2023-09-17 00:00:00              Part-Time   
..                   ...                  ...                    ...   
295  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
296  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
297  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
298  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   
299  2024-04-08 00:00:00  2025-03-18 00:00:00              Part-Time   

     COURSE FUNDING REGISTRATION FEE PAYMENT MODE  COURSE FEE  
0 



---

### Checking students which are not found in semester_results

In [10]:
# Get sets of Student IDs
ids_profiles = set(student_profiles['STUDENT ID'].dropna().unique())
ids_results = set(semester_results['STUDENT ID'].dropna().unique())

# Find mismatched IDs (present in student_profiles but not in student_profiles)
only_in_results =  ids_profiles - ids_results

# Filter and print those rows from student_profiles
mismatched_rows = student_profiles[student_profiles['STUDENT ID'].isin(only_in_results)]

print("❌ Mismatched rows from student_profiles:")
mismatched_rows.head(20)


❌ Mismatched rows from student_profiles:


STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
108  2101-111/001      F                Y       NaN   
109  2101-111/002      F          Y   NaN       NaN   
110  2101-111/003      F          Y   NaN       NaN   
111  2101-111/004      F          Y   NaN       NaN   
112  2101-111/005      F          Y   NaN       NaN   
113  2101-111/006      F          Y   NaN       NaN   
114  2101-111/007      F          Y   NaN       NaN   
115  2101-111/008      F          Y   NaN       NaN   
300  5113-009/001      F        NaN   NaN         Y   
301  5113-009/002      F          Y   NaN       NaN   
302  5113-009/003      F          Y   NaN       NaN   
303  5113-009/004      M        NaN     Y       NaN   
304  5113-009/005      F          Y   NaN       NaN   
305  5113-009/006      F        NaN     Y       NaN   
306  5113-009/007      F          Y   NaN       NaN   

    COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
108                    Malaysian  08/07/1995                Degree   
109                          NaN  08/09/1997           Certificate   
110                               19/06/1999               Diploma   
111                               28/09/2010           Certificate   
112                          NaN  19/10/1990           Certificate   
113                               10/04/1999           Certificate   
114                               24/05/2000           Certificate   
115                               12/04/2001           Certificate   
300                        India  14/11/1985                Degree   
301                          NaN  28/01/1993                Degree   
302                          NaN  06/03/1995                Degree   
303                  Philippines  09/02/1995                Degree   
304                          NaN  01/10/1993                Degree   
305                    Indonesia  30/09/1996                Degree   
306                          NaN  06/11/1996                Degree   

                 NAME OF QUALIFICATION AND INSTITUTION  \
108  Bachelor of Science (HRD), Universiti Teknolog...   
109  Certificate in Grammar & Writing Intermediate ...   
110  Diploma in Mechatronic Engineering, Nee Ann Po...   
111        Higher Nitec in Hospitality Operations, ITE   
112                                           O' level   
113  Higher Nitec in Business Studies (Service Mana...   
114                                          O' levels   
115                                          O' levels   
300  Bachelor of Arts in Business with Logistics an...   
301  Bachelor of Business (Business Administration)...   
302  Bachelor of Science in Hotel Administration (H...   
303  Bachelor of Science in Accounting and Finance ...   
304  Bachelor of Commerce (Accounting and Finance)/...   
305  Bachelor of Economics (Accounting)/\nUniversit...   
306  Degree of Computer Science, National Universit...   

    DATE ATTAINED HIGHEST QUALIFICATION  \
108                          2020-04-04   
109                          2021-10-04   
110                          2020-12-24   
111                          2021-09-24   
112                          2010-07-24   
113                          2012-04-24   
114                          2020-04-24   
115                          2022-05-24   
300                          2019-08-04   
301                          2020-10-04   
302                          2019-06-08   
303                          2020-08-05   
304                          2019-11-05   
305                          2018-05-05   
306                          2021-04-20   

                                  DESIGNATION    COMMENCEMENT DATE  \
108                                   Manager  2025-04-24 00:00:00   
109                           Admin Assistant  2025-04-24 00:00:00   
110                                HR Manager  2025-04-24 00:00:00   
111                           Admin Executive  2025-04-24 00:00:00   
112                           Admin E

Students that are included in student_profile dataset but not in semester_results are those who haven't completed the course yet

#### Gender
---

In [11]:
# Check unique values in the 'GENDER' column
unique_genders = student_profiles['GENDER'].value_counts().reset_index()
unique_genders.columns = ['Gender', 'Count']
print("Unique values in 'GENDER' column:")
display(unique_genders)

Unique values in 'GENDER' column:


Gender  Count
0      F    265
1      M     42

From looking at the unique values, it seemed to be clean.

#### SG CITIZEN, SG PR & FOREIGNER
---

In [12]:
import pandas as pd
import numpy as np

# Load the student_profiles DataFrame
student_profiles = pd.read_excel("https://github.com/swamhtetg90/DAVI-CA2/blob/main/DAVI%20CA2%20datasets%20and%20meta%20data/Student%20Profiles.xlsx?raw=true")

# Replace empty strings with NaN in the specified columns
cols_to_check = ['SG CITIZEN', 'SG PR', 'FOREIGNER']
for col in cols_to_check:
    student_profiles[col] = student_profiles[col].replace(r'^\s*$', np.nan, regex=True)

# Check if there is only one non-null value among 'SG CITIZEN', 'SG PR', and 'FOREIGNER' for each row
student_profiles['Residential_Status_Check'] = (student_profiles[['SG CITIZEN', 'SG PR', 'FOREIGNER']].notna().sum(axis=1) == 1)

# Print rows where the condition is False (i.e., not exactly one non-null value)
mismatched_residential_status = student_profiles[student_profiles['Residential_Status_Check'] == False]

if mismatched_residential_status.empty:
    print("Each student has exactly one residential status specified.")
else:
    print("The following rows have more or less than one residential status specified:")
    display(mismatched_residential_status[['STUDENT ID', 'SG CITIZEN', 'SG PR', 'FOREIGNER', 'Residential_Status_Check']])

# Drop the temporary check column
student_profiles = student_profiles.drop(columns=['Residential_Status_Check'])

Each student has exactly one residential status specified.


Therefore, we can combine these 3 columns into 1 columnn called "Residential Status".

#### Nationality
---

In [13]:
# Check unique values in the 'COUNTRY OF OTHER NATIONALITY' column
unique_nationalities = student_profiles['COUNTRY OF OTHER NATIONALITY'].value_counts().reset_index()
unique_nationalities.columns = ['Nationality', 'Count']
print("Unique values in 'COUNTRY OF OTHER NATIONALITY' column:")
display(unique_nationalities)

Unique values in 'COUNTRY OF OTHER NATIONALITY' column:


Nationality  Count
0                  88
1     Malaysia     27
2        China     14
3        India     12
4  Philippines      8
5      Myanmar      3
6  Netherlands      1
7    Malaysian      1
8      Vietnam      1
9    Indonesia      1

In [14]:
# Count the occurrences of each value in the 'SG CITIZEN' column
sg_citizen_counts = student_profiles['SG CITIZEN'].value_counts().reset_index()

# Rename columns for clarity
sg_citizen_counts.columns = ['SG CITIZEN Value', 'Count']

# Display the counts
print("Count of each value in the 'SG CITIZEN' column:")
display(sg_citizen_counts)

Count of each value in the 'SG CITIZEN' column:


SG CITIZEN Value  Count
0                Y    224
1              Yes     15

Based on the results, you can see that if it is blank in "COUNTRY OF OTHER NATIONALITY", the Nationality is Singapore

#### DOB (Date of Birth)
---

In [15]:
# Check for missing values in the 'DOB' column
missing_dob = student_profiles['DOB'].isnull().sum()
print(f"Number of missing values in 'DOB': {missing_dob}")

# Attempt to convert 'DOB' to datetime, coercing errors
student_profiles['DOB_datetime'] = pd.to_datetime(student_profiles['DOB'], errors='coerce', format='%d/%m/%Y')

# Check for values that couldn't be parsed (will be NaT - Not a Time)
invalid_dob = student_profiles[student_profiles['DOB_datetime'].isna() & student_profiles['DOB'].notna()]

print("\nRows with invalid 'DOB' values:")
display(invalid_dob[['STUDENT ID', 'DOB', 'DOB_datetime']])

# Drop the temporary datetime column
student_profiles = student_profiles.drop(columns=['DOB_datetime'])

Number of missing values in 'DOB': 0

Rows with invalid 'DOB' values:


STUDENT ID          DOB DOB_datetime
105   2101-110/006  13-Feb-1984          NaT
106   2101-110/007  13-Jul-1987          NaT
107   2101-110/008  15-Jul-1994          NaT
184  2102-067A/011   16-07-1991          NaT

From looking at values that cannot be chnged to datetime, it shows that month column is in its short abbreviation form instead of number.

So, we will have to convert it.

#### Highest Qualification
----

In [16]:
# Check unique values in the 'HIGHEST QUALIFICATION' column
unique_qualifications = student_profiles['HIGHEST QUALIFICATION'].value_counts().reset_index()
unique_qualifications.columns = ['Highest Qualification', 'Count']
print("Unique values in 'HIGHEST QUALIFICATION' column:")
display(unique_qualifications)

Unique values in 'HIGHEST QUALIFICATION' column:


Highest Qualification  Count
0                Degree    137
1           Certificate     87
2               Diploma     71
3                Master     11
4                            1

In [17]:
# Check for rows where 'HIGHEST QUALIFICATION' is null or empty
null_qualification_rows = student_profiles[student_profiles['HIGHEST QUALIFICATION'].isnull() | (student_profiles['HIGHEST QUALIFICATION'] == ' ') | (student_profiles['HIGHEST QUALIFICATION'] == '')]

print("Rows with null or empty 'HIGHEST QUALIFICATION':")
display(null_qualification_rows)

Rows with null or empty 'HIGHEST QUALIFICATION':


STUDENT ID GENDER SG CITIZEN SG PR FOREIGNER  \
128  2102-063/013      F        NaN   NaN         Y   

    COUNTRY OF OTHER NATIONALITY         DOB HIGHEST QUALIFICATION  \
128                      Vietnam  24/07/1989                         

    NAME OF QUALIFICATION AND INSTITUTION DATE ATTAINED HIGHEST QUALIFICATION  \
128                                                                2016-06-06   

            DESIGNATION    COMMENCEMENT DATE      COMPLETION DATE  \
128  Accounts Executive  2022-04-18 00:00:00  2022-09-14 00:00:00   

    FULL-TIME OR PART-TIME COURSE FUNDING REGISTRATION FEE PAYMENT MODE  \
128              Part-Time     Individual              107         Nets   

     COURSE FEE  
128         888

Since only 1 row have got null values for  Highest Qualification, we will drop it.

#### Name of Qualification and Institution
---

In [18]:
# Split 'NAME OF QUALIFICATION AND INSTITUTION' into 'QUALIFICATION_NAME' and 'INSTITUTION_NAME'
# Use regex to split on either comma or slash
student_profiles[['QUALIFICATION_NAME', 'INSTITUTION_NAME']] = student_profiles['NAME OF QUALIFICATION AND INSTITUTION'].str.split(r',|/', n=1, expand=True)

# Display the new columns and the original column to verify
display(student_profiles[['NAME OF QUALIFICATION AND INSTITUTION', 'QUALIFICATION_NAME', 'INSTITUTION_NAME']])

NAME OF QUALIFICATION AND INSTITUTION  \
0                                                  SPM   
1                    Certificate in Office Skills, ITE   
2    Bachelor of Business Administration, Universit...   
3    Office Management Diploma, NCOI Rotterdam, The...   
4    Diploma in Business Admininstration, LCCI Leve...   
..                                                 ...   
302  Bachelor of Science in Hotel Administration (H...   
303  Bachelor of Science in Accounting and Finance ...   
304  Bachelor of Commerce (Accounting and Finance)/...   
305  Bachelor of Economics (Accounting)/\nUniversit...   
306  Degree of Computer Science, National Universit...   

                                    QUALIFICATION_NAME  \
0                                                  SPM   
1                         Certificate in Office Skills   
2                  Bachelor of Business Administration   
3                            Office Management Diploma   
4                  Diploma in Business Admininstration   
..                                                 ...   
302  Bachelor of Science in Hotel Administration (H...   
303  Bachelor of Science in Accounting and Finance ...   
304      Bachelor of Commerce (Accounting and Finance)   
305                 Bachelor of Economics (Accounting)   
306                         Degree of Computer Science   

                                     INSTITUTION_NAME  
0                                                None  
1                                                 ITE  
2                      University of Rajasthan, India  
3                     NCOI Rotterdam, The Netherlands  
4     LCCI Level 3, Private Secretary's Diploma, LCCI  
..                                                ...  
302                 \nUniversity of Nevada, Las Vegas  
303                            \nUniversity of London  
304                 \nCurtin University of Technology  
305                   \nUniversitas Katolik Indonesia  
306                  National University of Singapore  

[307 rows x 3 columns]

In [19]:
# Check unique values in the 'QUALIFICATION_NAME' column
unique_qualification_names = student_profiles['QUALIFICATION_NAME'].value_counts().reset_index()
unique_qualification_names.columns = ['Qualification Name', 'Count']
print("Unique values in 'QUALIFICATION_NAME' column:")
display(unique_qualification_names) # Displaying head as there might be many unique values

# Check unique values in the 'INSTITUTION_NAME' column
unique_institution_names = student_profiles['INSTITUTION_NAME'].value_counts().reset_index()
unique_institution_names.columns = ['Institution Name', 'Count']
print("\nUnique values in 'INSTITUTION_NAME' column:")
display(unique_institution_names) # Displaying head as there might be many unique values

Unique values in 'QUALIFICATION_NAME' column:


Qualification Name  Count
0                                                  SPM     15
1                                             O' level     15
2                   Diploma in Business Administration      7
3                  Bachelor of Business Administration      6
4                                            O' levels      6
..                                                 ...    ...
171                   Bachelor of Business (Marketing)      1
172  Bachelor of Commerce in Management and Marketi...      1
173                       Bachelor of Arts (Sociology)      1
174                      Bachelor of Arts with Honours      1
175                         Degree of Computer Science      1

[176 rows x 2 columns]


Unique values in 'INSTITUTION_NAME' column:


Institution Name  Count
0                                      ITE     13
1       \nNational University of Singapore     12
2                     \nMurdoch University      8
3                              PSB Academy      6
4                    Singapore Polytechnic      6
..                                     ...    ...
129                  \nUniversity of Delhi      1
130                 \nUniversity of Malaya      1
131    \nNational Technological University      1
132                  \nNanyang Polytechnic      1
133  \nQueensland University of Technology      1

[134 rows x 2 columns]

#### DATE ATTAINED HIGHEST QUALIFICATION
----

In [33]:
# Convert 'DATE ATTAINED HIGHEST QUALIFICATION' to datetime objects
student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'] = pd.to_datetime(student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'], errors='coerce')

# Check for unrealistic dates: Date attained should be after DOB and likely before or around COMMENCEMENT DATE
unrealistic_qualification_dates = student_profiles[
    (student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'].notna()) &
    (
        (student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'] < student_profiles['DOB']) |
        (student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'] > student_profiles['COMMENCEMENT DATE']) # Assuming qualification is attained before or around course start
    )
]

print("Rows with potentially unrealistic 'DATE ATTAINED HIGHEST QUALIFICATION':")
display(unrealistic_qualification_dates[['STUDENT ID', 'DOB', 'DATE ATTAINED HIGHEST QUALIFICATION', 'COMMENCEMENT DATE']])

# Check for missing values after conversion
missing_qualification_dates = student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'].isna().sum()
print(f"\nNumber of missing values in 'DATE ATTAINED HIGHEST QUALIFICATION' after conversion: {missing_qualification_dates}")

Rows with potentially unrealistic 'DATE ATTAINED HIGHEST QUALIFICATION':


Empty DataFrame
Columns: [STUDENT ID, DOB, DATE ATTAINED HIGHEST QUALIFICATION, COMMENCEMENT DATE]
Index: []


Number of missing values in 'DATE ATTAINED HIGHEST QUALIFICATION' after conversion: 0


In [40]:
# Convert 'DOB' to datetime objects, coercing errors
student_profiles['DOB'] = pd.to_datetime(student_profiles['DOB'], errors='coerce', format='%d/%m/%Y')

# Convert 'DATE ATTAINED HIGHEST QUALIFICATION' to datetime, if not already
student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'] = pd.to_datetime(
    student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'], errors='coerce', format='%d/%m/%Y')

# Calculate age in years at time of qualification
student_profiles['QUALIFICATION_AGE_YEARS'] = (
    (student_profiles['DATE ATTAINED HIGHEST QUALIFICATION'] - student_profiles['DOB']).dt.days / 365.25
).round(1)

# Sort the DataFrame by calculated age in ascending order
sorted_qualification_age = student_profiles.sort_values(by='QUALIFICATION_AGE_YEARS', ascending=True)

print("Age in years at the time of attaining highest qualification (sorted ascending):")
display(sorted_qualification_age[['STUDENT ID', 'DOB', 'NAME OF QUALIFICATION AND INSTITUTION', 'DATE ATTAINED HIGHEST QUALIFICATION', 'QUALIFICATION_AGE_YEARS']])

Age in years at the time of attaining highest qualification (sorted ascending):


STUDENT ID        DOB  \
111   2101-111/004 2010-09-28   
33    1101-012/001 1988-12-04   
113   2101-111/006 1999-04-10   
8     1101-009/009 1983-11-30   
12    1101-010/002 1973-04-20   
..             ...        ...   
195   2102-069/001 1961-04-01   
105   2101-110/006        NaT   
106   2101-110/007        NaT   
107   2101-110/008        NaT   
184  2102-067A/011        NaT   

                 NAME OF QUALIFICATION AND INSTITUTION  \
111        Higher Nitec in Hospitality Operations, ITE   
33               Nitec in Service Skills (Office), ITE   
113  Higher Nitec in Business Studies (Service Mana...   
8                                             N' Level   
12                                            O' level   
..                                                 ...   
195       O level, Private Secretary Certificate, LCCI   
105  Bachelor of Commerce (Management and Marketing...   
106              Bachelor of Science in Business, SUSS   
107  Diploma in Integrated Events Management, Repub...   
184  Bachelor of Science (Nursing), Curtin Universi...   

    DATE ATTAINED HIGHEST QUALIFICATION  QUALIFICATION_AGE_YEARS  
111                          2021-09-24                     11.0  
33                           2001-11-06                     12.9  
113                          2012-04-24                     13.0  
8                            2000-05-07                     16.4  
12                           1990-02-09                     16.8  
..                                  ...                      ...  
195                          2019-02-28                     57.9  
105                          2007-05-16                      NaN  
106                          2010-01-16                      NaN  
107                          2018-09-16                      NaN  
184                          2019-02-22                      NaN  

[307 rows x 5 columns]

From looking at this data, I realized that it is impossible for some of these students to have these qualification. Therefore, we will be dropping them.
`[ 111, 33, 113, 38, 267]`

#### Designation
---

In [31]:
# Check unique values and their counts in the 'DESIGNATION' column
designation_counts = student_profiles['DESIGNATION'].value_counts().reset_index()
designation_counts.columns = ['Designation', 'Count']

print("Unique values and counts in 'DESIGNATION' column:")
# Display top 50 designations if there are many unique values
if len(designation_counts) > 50:
    display(designation_counts.head(50))
    print(f"\n... and {len(designation_counts) - 50} more unique designations.")
else:
    display(designation_counts)

Unique values and counts in 'DESIGNATION' column:


Designation  Count
0                                          -     35
1                            Admin Assistant     14
2                               HR Executive     13
3                            Admin Executive      8
4                                    Manager      8
5                                 HR Manager      8
6                                  Secretary      6
7                               HR Assistant      6
8             Business Development Executive      4
9                        Senior HR Executive      4
10                        Accounts Executive      3
11                          HR Administrator      3
12                            Senior Manager      3
13                                      N.A.      3
14                             Admin Manager      3
15                          Admin Supervisor      3
16                                 Executive      3
17                             Admin Officer      3
18                             Administrator      3
19                              IT Developer      3
20                       Marketing Executive      2
21                  Human Resource Assistant      2
22                                Accountant      2
23                                Consultant      2
24                        HR & Admin Officer      2
25                Senior Recruitment Manager      2
26                          Snr HR Executive      2
27                      Admin & HR Assistant      2
28                       Executive Assistant      2
29            Early Childhood Support Office      2
30                          Senior Executive      2
31                 Executive, Administration      2
32                                     Clerk      2
33                                HR & Admin      2
34                 Admin & Finance Executive      2
35                       Operation Executive      2
36                             HR Specialist      2
37                           Audit Associate      2
38                    Recruitment Consultant      2
39                                   Teacher      2
40                  Assistant Vice-President      2
41                                  Lecturer      2
42                                Admin Lead      2
43            Human Resource & Admin Manager      2
44                  Human Resource Executive      2
45                  Administrative Executive      2
46                                HR Officer      2
47  Senior Assistant (Tenant Administration)      1
48                            Office Manager      1
49               HR Support / Office Manager      1


... and 114 more unique designations.


In [30]:
from collections import Counter
import re

# Combine all designations into a single string, handling potential NaNs
all_designations = ' '.join(student_profiles['DESIGNATION'].dropna().astype(str).str.lower())

# Split the string into words, using regex to find word characters
words = re.findall(r'\b\w+\b', all_designations)

# Count the frequency of each word
word_counts = Counter(words)

# Convert to a DataFrame for easier display
word_counts_df = pd.DataFrame.from_dict(word_counts, orient='index', columns=['Count']).reset_index()
word_counts_df = word_counts_df.rename(columns={'index': 'Word'})

# Sort by count in descending order
word_counts_df = word_counts_df.sort_values(by='Count', ascending=False)

print("Count of each word in 'DESIGNATION' column:")
display(word_counts_df.head(50)) # Displaying the top 50 most frequent words

Count of each word in 'DESIGNATION' column:


Word  Count
6         executive     81
1                hr     65
0             admin     57
5           manager     47
2         assistant     41
19           senior     21
16            human     14
21          officer     14
17         resource     11
8     administrator      9
14         business      9
25        associate      9
3           support      7
29              snr      7
15      development      6
57       specialist      6
18        secretary      6
69          payroll      6
4            office      5
31         accounts      5
63       consultant      5
24        marketing      5
7    administration      4
33        developer      4
9           finance      4
26           client      4
27         services      4
68             lead      4
90      recruitment      4
81       supervisor      4
72      coordinator      4
87       operations      4
93             head      4
43   administrative      3
51          service      3
64             vice      3
10        operation      3
32         software      3
39         division      3
73       accountant      3
76          analyst      3
98                a      3
55             team      3
65        president      3
125              it      3
97                n      3
99         director      3
13        childhood      2
12            early      2
56             apac      2

In [32]:
# Function to categorize designations based on keywords
def categorize_designation(designation):
    if pd.isna(designation) or designation.strip() in ['', '-', 'N.A.']:
        return 'Unknown'
    designation = str(designation).lower()
    if 'manager' in designation:
        return 'Manager'
    elif 'executive' in designation:
        return 'Executive'
    elif 'assistant' in designation:
        return 'Assistant'
    elif 'officer' in designation:
        return 'Officer'
    elif 'specialist' in designation:
        return 'Specialist'
    elif 'consultant' in designation:
        return 'Consultant'
    elif 'coordinator' in designation:
        return 'Coordinator'
    elif 'head' in designation:
        return 'Head'
    elif 'director' in designation:
        return 'Director'
    elif 'analyst' in designation:
        return 'Analyst'
    elif 'administrator' in designation:
        return 'Administrator'
    elif 'clerk' in designation:
        return 'Clerk'
    elif 'teacher' in designation or 'lecturer' in designation:
        return 'Educator'
    elif 'audit' in designation:
        return 'Audit'
    elif 'finance' in designation or 'accountant' in designation:
        return 'Finance/Accounting'
    elif 'marketing' in designation or 'business development' in designation:
        return 'Marketing/BD'
    elif 'recruitment' in designation or 'hr' in designation or 'human resource' in designation:
        return 'HR/Recruitment'
    elif 'it' in designation or 'developer' in designation:
        return 'IT/Tech'
    elif 'operation' in designation:
        return 'Operations'
    # Add more categories as needed
    else:
        return 'Other'

# Apply the categorization function
student_profiles['DESIGNATION_CATEGORY'] = student_profiles['DESIGNATION'].apply(categorize_designation)

# Check the counts of the new categories
designation_category_counts = student_profiles['DESIGNATION_CATEGORY'].value_counts().reset_index()
designation_category_counts.columns = ['Designation Category', 'Count']
print("\nCounts of Designation Categories:")
display(designation_category_counts)


Counts of Designation Categories:


Designation Category  Count
0             Executive     81
1               Manager     47
2               Unknown     38
3                 Other     36
4             Assistant     32
5               Officer     14
6         Administrator      9
7        HR/Recruitment      9
8               IT/Tech      7
9            Specialist      6
10           Consultant      5
11             Educator      4
12          Coordinator      4
13                 Head      4
14              Analyst      3
15   Finance/Accounting      3
16                Clerk      2
17                Audit      2
18             Director      1

#### Commence Date & Completion Date
---

In [23]:
# Convert 'COMMENCEMENT DATE' and 'COMPLETION DATE' to datetime objects
student_profiles['COMMENCEMENT DATE'] = pd.to_datetime(student_profiles['COMMENCEMENT DATE'], errors='coerce')
student_profiles['COMPLETION DATE'] = pd.to_datetime(student_profiles['COMPLETION DATE'], errors='coerce')

# Add 'COURSE_COMPLETED' column
student_profiles['COURSE_COMPLETED'] = student_profiles['COMPLETION DATE'].notna()

# Display the relevant columns to verify
display(student_profiles[['STUDENT ID', 'COMMENCEMENT DATE', 'COMPLETION DATE', 'COURSE_COMPLETED']])

STUDENT ID COMMENCEMENT DATE COMPLETION DATE  COURSE_COMPLETED
0    1101-009/001        2022-04-18      2023-09-17              True
1    1101-009/002        2022-04-18      2023-09-17              True
2    1101-009/003        2022-04-18      2023-09-17              True
3    1101-009/004        2022-04-18      2023-09-17              True
4    1101-009/005        2022-04-18      2023-09-17              True
..            ...               ...             ...               ...
302  5113-009/003        2025-04-14             NaT             False
303  5113-009/004        2025-04-14             NaT             False
304  5113-009/005        2025-04-14             NaT             False
305  5113-009/006        2025-04-14             NaT             False
306  5113-009/007        2025-04-14             NaT             False

[307 rows x 4 columns]

In [26]:
# Get the student IDs from rows with invalid commencement dates
student_ids_with_invalid_commencement = invalid_commencement_dates['STUDENT ID']

# Check which of these student IDs exist in the semester_results DataFrame
students_in_semester_results_bool = student_ids_with_invalid_commencement.isin(semester_results['STUDENT ID'])

# Count students with invalid commencement dates who are in semester_results
count_in_semester_results = students_in_semester_results_bool.sum()

# Count students with invalid commencement dates who are not in semester_results
count_not_in_semester_results = (~students_in_semester_results_bool).sum()

print(f"Number of students with invalid commencement dates found in semester_results: {count_in_semester_results}")
print(f"Number of students with invalid commencement dates not found in semester_results: {count_not_in_semester_results}")

# Display the rows from semester_results for students with invalid commencement dates (as done before)
students_in_semester_results_df = semester_results[semester_results['STUDENT ID'].isin(student_ids_with_invalid_commencement)]
print("\nRows from semester_results for students with invalid commencement dates:")
display(students_in_semester_results_df)

Number of students with invalid commencement dates found in semester_results: 5
Number of students with invalid commencement dates not found in semester_results: 0

Rows from semester_results for students with invalid commencement dates:


STUDENT ID PERIOD  GPA
15   1101-009/006  Sem 1  2.1
16   1101-009/006  Sem 2  2.2
17   1101-009/006  Sem 3  2.3
174  5113-005/005  Sem 1  3.1
175  5113-005/005  Sem 2  3.5
201  1101-011/008  Sem 1  3.3
202  1101-011/008  Sem 2  3.8
203  1101-011/008  Sem 3  3.4
346  2102-064/013   Sem1  3.6
496  2102-069/010  Sem 1  3.5
508  2102-069/010  Sem 1  3.5

From looking at this, i can insert the commence date and and completion date based on rows that have the same `COURSE_ID` and `INTAKE_NO`

#### FULL-TIME OR PART-TIME
----

In [28]:
# Check unique values in the 'FULL-TIME OR PART-TIME' column
unique_course_type = student_profiles['FULL-TIME OR PART-TIME'].value_counts().reset_index()
unique_course_type.columns = ['Course Type', 'Count']
print("Unique values in 'FULL-TIME OR PART-TIME' column:")
display(unique_course_type)

Unique values in 'FULL-TIME OR PART-TIME' column:


Course Type  Count
0   Part-Time    237
1   Full-Time     41
2   Part Time     29

Based on the count, I believe we can combine `Part-Time` and `Part Time` together by changing ` ` to `-`

## Data Cleaning
---

### Data Cleaning Steps for Student Profile

#### 1. Student ID Decomposition

* Split `STUDENT_ID` into three new columns:

  * `COURSE_ID`
  * `INTAKE_NO`
  * `INDEX_NO`
* Format: `<COURSE_ID>-<INTAKE_NO>/<INDEX_NO>`

#### 2. Remove Unmatched Semester Results

* Remove entries from `semester_results` if the student does not exist in the `student profile` table.

#### 3. Gender

* Clean - Have 2 unique values ('M' and 'F')

#### 4. Residential Status

* Combine `SG CITIZEN`, `SG PR`, and `FOREIGNER` into a single column: `RESIDENTIAL_STATUS`.

  * Logic: Only one of the three columns contains "Y"; others are null.
  * New values: `Singapore Citizen`, `PR`, or `Foreigner`.

#### 5. Nationality

* Rename `COUNTRY OF OTHER NATIONALITY` to `NATIONALITY`.
* Standardize values:
  * Correct spelling errors (e.g., `Malaysian` → `Malaysia`)
* Replace null or blank values with: `Singapore`.

#### 6. Date of Birth (`DOB`)

* Not in datetime format
* Some rows are in short abbreviation for month instead of numbers

#### 7. Highest Qualification

* Drop rows where value is `" "`

#### 8. Name of Qualification and Institution

* Split `NAME OF QUALIFICATION AND INSTITUTION` into two parts:

  * `QUALIFICATION_NAME` (text before the first comma)
  * `INSTITUTION_NAME` (text after the first comma)

#### 9. Date Attained Highest Qualification

* Check if the date is realistic compared to `DOB` and `COMMENCE DATE`
* Remove these rows `[ 111, 33, 113, 38, 267]`
  * These shows students which doesn't meet the requirement's age for their Qualification.

#### 10. Designation

* Key-Word Mapping to reduce unique values

#### 11. Commence Date & Completion Date

* Convert to datetime format
* Insert blank values by looking at `COURSE_NO` & `INTAKE_NO`

#### 12. Full-time or Part-time

* Standardize values:
  * Correct spelling errors (e.g., `Full Time` → `Full-Time`).

#### 13. Course Funding Type

* Standardize values to one of the following:

  * `Individual`
  * `Individual - SFC (SkillsFuture Credit)`
  * `Sponsored`
  * `Individual - Waived App Fee`
* Check for spelling errors and inconsistencies.

#### 14. Registration Fee

* Remove `$` symbol.
* Convert to float with 2 decimal places.

#### 15. Payment Mode

* Standardize text and check for spelling mistakes.

#### 16. Course Fee

* Remove `$` symbol.
* Convert to float with 2 decimal places.

---


## Exporting Data
---